crates/ark/src/lsp/diagnostics.rs

```diff
@@ -780,13 +807,116 @@ fn recurse_call(
+            }
+        },
+        _ => {},
     };
 
+    // Continue with default recursion to handle any other arguments
+    recurse_call_like_arguments_default(node, context, diagnostics)?;
+
     ().ok()
 }
 
+fn handle_package_attach_call(node: Node, context: &mut DiagnosticContext) -> anyhow::Result<()> {
+    // Find the first argument (package name). Positionally for now, no attempt
+    // at argument matching whatsoever.
+    let Some(package_node) = node.arguments_values().nth(0) else {
+        return Err(anyhow::anyhow!("Can't unpack attached package argument"));
+    };
+
+    // Just bail if `character.only` is passed, even if it's actually `FALSE`.
+    // We'll do better when we have a more capable argument inspection
+    // infrastructure.
+    if let Some(_) = node
+        .arguments_names_as_string(context.contents)
+        .find(|n| n == "character.only")
+    {
+        return Ok(());
+    }
+
+    let package_name = package_node.get_identifier_or_string_text(context.contents)?;
+    let attach_pos = node.end_position();
+
+    let package = insert_package_exports(&package_name, attach_pos, context)?;
+
+    // Also attach packages from `Depends` field
+    let mut attach_dependencies = package.description.depends.clone();
+
+    // Special handling for the tidyverse and tidymodels packages. Hard-coded
+    // for now but in the future, this should probably be expressed as a
+    // `DESCRIPTION` field like `Config/Needs/attach`.
+    let attach_field = match package.description.name.as_str() {
+        // https://github.com/tidyverse/tidyverse/blob/0231aafb/R/attach.R#L1
+        "tidyverse" => {
+            vec![
+                "dplyr",
+                "readr",
+                "forcats",
+                "stringr",
+                "ggplot2",
+                "tibble",
+                "lubridate",
+                "tidyr",
+                "purrr",
+            ]
+        },
+        // https://github.com/tidymodels/tidymodels/blob/aa3f82cf/R/attach.R#L1
+        "tidymodels" => {
+            vec![
+                "broom",
+                "dials",
+                "dplyr",
+                "ggplot2",
+                "infer",
+                "modeldata",
+                "parsnip",
+                "purrr",
+                "recipes",
+                "rsample",
+                "tibble",
+                "tidyr",
+                "tune",
+                "workflows",
+                "workflowsets",
+                "yardstick",
+            ]
+        },
+        _ => vec![],
+    };
+    attach_dependencies.extend(attach_field.into_iter().map(String::from));
+
+    for package_name in attach_dependencies {
+        insert_package_exports(&package_name, attach_pos, context)?;
+    }
+
+    Ok(())
+}
+
+fn insert_package_exports(
+    package_name: &str,
+    attach_pos: Point,
+    context: &mut DiagnosticContext,
+) -> anyhow::Result<Arc<Package>> {
```

Comment:
Yea I'm fairly certain we can and should just return `anyhow::Result<&Package>`? And that `get()` method really should return a reference, not an `Arc`.

We'd still clone out the `package.namespace.exports`, which is fine and seems required, but otherwise `Package` is immutable

Reply:
Can't do that for the reason mentioned in the other comment (sharing across threads).

Just to be clear, `Package` _is_ immutable. An `Arc` confers shared ownership in the sense that it's only deallocated when all instances go out of scope. But the ownership is not in terms of being able to mutate it. Essentially all instances are like non-mutable references that you can share across threads without lifetime issues.